---
title: "Seurat_Code_QC_2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

Begin by installing and loading necessary packages. Unhash the "install" script if you need to install them. You only need to install once per machine, unless you want to update. 

```{r}
#install.packages("BiocManager")
#BiocManager::install("Seurat")
#BiocManager::install("ggplot2")
#install.packages("dplr")
#install.packages("ape")
#install.packages("cowplot")
#install.packages("Matrix")
```

Load packages into your workspace. 

```{r}
library(Seurat)
library(ggplot2)
library(dplyr)
library(ape)
library(cowplot)
library(Matrix)

#find out where you are
getwd()
```

*Edit the following code every time* and make sure the folders "QC" and ... are present in the "dir" folder above
```{r}
#Specify your preferred directory for all input + output
dir= "/Users/whippoorwill/Desktop/hack/"

#Specify exactly where your matrix files/raw data live
datafolder = "task2.gene_expr"


#Add in any folder name within each sample that leads to your data files: 
innerfolder = NULL

#Specify what kind of data you have: "CellRanger" Output (barcodes + matrix .tsv files) or "counts" data (usually something ending in counts.tsv)
datatype = "CellRanger"

#if "datatype" = "counts", specify "filename", otherwise "NA" or "NULL"
filename = "GEX_counts.tsv.gz"

#Set minimums for genes/cell (min.features) and cell expressing a given gene (min.cells). Default is 1/1, you can filter later. If the object is very large, filter here. 
min.cells = 1
min.features = 1

#Name your project something descriptive
project<-"hackathon2021-task2" 

#Specify your organism; please capitalize the first letter (i.e. "Mouse", "Human","Zebrafish","Rat")
organism = "Human"

#Add in metadata. You may want to run the next chunk and then return to edit this so it is in the same order as "ids"
#age = c(rep("P7",2))
#condition = c("Deprived","Control")

#metrics you want to look at for QC
m = c("nCount_RNA","nFeature_RNA","percent.mito")

#set up folders
QCfolder = "QC"
```

*Please note you will also have to edit the thresholding chunk below, line 258. You may run everything up until then*

Then create an object with the folder names that you can use to open each sample folder when you need it. The following code will allow you to open all the folders in sequence. 
```{r}
ids = list.files(file.path(dir,datafolder))
ids
```


For each ID (i.e. folder), read in the data and change the colomn names to look like: "Barcode - id". If the second chunk doesn't work, try manually unzipping the file. Make sure to change the 'filename' extension above. 
```{r}
if (datatype == "CellRanger"){
  d10x.data = sapply(ids, function(id){
    d10x = Read10X(file.path(dir,datafolder,id)) #read in each matrix
    d10x
  })
}

if (datatype == "counts"){
  d10x.data = sapply(ids, function(id){
    d10x =as.sparse(read.table(file.path(dir,datafolder,id,filename),sep = '\t',header = T,row.names = 1))
    colnames(d10x) = paste(sapply(strsplit(colnames(d10x),split="-"),'[[',1L),id,sep="-")
    d10x
  })
}

experiment.data<-do.call("cbind",d10x.data)
```

The d10x.data object is a list of arrays whose columns consist of cell IDs and whose rows represent genes. 
Make experiment.data, a single array with as many columns as total cells. 
****
"Aggregate" the data which cellRanger can do, but not well. Create a seurat object, keeping genes that are in >3 cells, cells with >50 genes (you can always set these to zero and filter later)

"names.field" and "names.delim" determines the input identity for each sample. IE if the cells are named "barcode - id", the "name" for each sample (listed in orig.ident) will be id.  

```{r}
sobject <- CreateSeuratObject(
        experiment.data,
        project = project, 
        min.cells = min.cells, 
        min.features = min.features
        #names.field = 2, #put the "id" from above into meta data
        #names.delim = "\\-"
        )

table(sobject$orig.ident)
```

Calculate mitochondrial and ribosomal content, add as metadata to each cell (column). Percent mito is literally dividing one matrix by another matrix to get a percentage. Grep searches through rownames of the data (i.e. genes) for anything that starts with mt. 

```{r}
if (organism %in% c("Mouse","Rat")){
  sobject <- PercentageFeatureSet(sobject, pattern = "^mt-", col.name = "percent.mito")
  sobject <- PercentageFeatureSet(sobject, pattern = "^Rp", col.name = "percent.ribo")
}
if (organism == "Human"){
  sobject <- PercentageFeatureSet(sobject, pattern = "^MT-", col.name = "percent.mito")
  sobject <- PercentageFeatureSet(sobject, pattern = "^RP", col.name = "percent.ribo")
}
if (organism == "Zebrafish"){
  sobject <- PercentageFeatureSet(sobject, pattern = "^mt-", col.name = "percent.mito")
  sobject <- PercentageFeatureSet(sobject, pattern = "^rp", col.name = "percent.ribo")
}
```
***
Note that the metadata column "orig.ident" always holds your sample IDs (i.e. the names of the data folders). Don't overwrite it. 

Add sample descriptions as metadata columns.
**Skip***
```{r}
samplename = as.character(sobject$orig.ident)
names(samplename) = names(sobject$orig.ident)

#Create or read in a metadata file
sample_description = paste0(condition,"-",age)
meta = as.data.frame(cbind(age,condition,sample_description))
rownames(meta) = ids
meta

#replace the names with any column of "meta" - add anything to meta and the following if necessary
#Pull out the item within meta that corresponds to each samplename(orig.ident) and the correct colum (age/condition/etc.)
age = sapply(samplename,function(x)meta[x,"age"])
condition = sapply(samplename,function(x)meta[x,"condition"])
sample_description = sapply(samplename,function(x)meta[x,"sample_description"])

#add new metadata columns
sobject$sample_description = sample_description
sobject$age = age
sobject$condition = condition

#check
table(sobject$sample_description,sobject$orig.ident)
table(sobject$age,sobject$condition)
table(sobject$age,sobject$sample_description)
```

For mouse/human cells, you can determine from Xist expression which ones are male and which are female. 
```{r}

#identify cell names with Xist expression


malegenes = GetAssayData(sobject,slot = "counts")
x = malegenes[c("DDX3Y"),]

sobject$malegenes = GetAssayData(sobject,slot = "counts")[c("DDX3Y"),]
sobject$femalegenes = colSums(GetAssayData(sobject,slot = "counts")[c("XIST","TSIX"),])

female<-WhichCells(object = sobject, cells = colnames(sobject)[sobject$femalegenes > 0 & sobject$malegenes ==0],slot = "counts") 

male<-WhichCells(object = sobject, cells = colnames(sobject)[sobject$femalegenes == 0 & sobject$malegenes > 0],slot = "counts") 
length(female)
length(male)


mixed = WhichCells(object = sobject, cells = colnames(sobject)[sobject$femalegenes > 0 & sobject$malegenes > 0],slot = "counts") 
neither = WhichCells(object = sobject, cells = colnames(sobject)[sobject$femalegenes == 0 & sobject$malegenes == 0],slot = "counts") 


#set up a vector with the correct number of "male" and "female" labels
sex=as.factor(c(rep("Female",length(female)),rep("Male",length(male)),rep("mixed",length(mixed)),rep("neither",length(neither)))) 

#assign each cell ID to its appropriate label; must match order set up in the vector above.
names(sex)<-c(female,male,mixed,neither) 
table(sex)

#order this vector by the cellnames vector found in names(DG@meta.data)
cellnames = rownames(sobject@meta.data)
sex = sex[match(cellnames,names(sex))]
sex = as.character(sex)
sex[is.na(sex)] = "Unknown"

table(sex)
#Add in a metadata column for the sex of the cell/animal
sobject$sexstringent <- sex

VlnPlot(sobject,features = c("TSIX","XIST","DDX3Y"),group.by = "sexstringent")
```
***
*QC - decide on final filtering parameters, look for clear outliers/differences between samples.*

1. Function to make an annotated violin plot
```{r}
anno_vln = function(seurat_object = sobject, metric = "nCount_RNA", group = "orig.ident"){
  #specify number of categories
  n = length(levels(seurat_object@meta.data[,group]))

  #make a violin plot
  a = VlnPlot(seurat_object,metric,group.by = group,pt.size = 0,log = F,cols = rep("lightgrey",n))

  #make a simple function that quantifies percentiles for each column
  add.percentiles = function(x) {
    r = quantile(x, probs = seq(0,1,0.2))
    r
  }

  a + geom_boxplot(width = 0.05,outlier.size = 0,notchwidth = 10,notch = T) + theme_bw() + stat_summary(fun=add.percentiles, geom="point", pch="_", colour=c(rep(c("red3","purple","yellow","forestgreen","darkblue","grey"),n)), size=10)
}
```

Plot each metric. Red = 0%, Purple = 20%, Yellow = 40%, Black = 50%,Green = 60%, Blue = 80%, Grey = 100% percentile
```{r}
#Save graphs
for (i in m){
  graph = anno_vln(metric = i)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(i,project,".eps")))
  print(graph)
  dev.off()
}

#print graphs to console
for (i in m){
  graph = anno_vln(metric = i)
  print(graph)
}
```

*****
Plot the number of cells each gene is represented by - i.e. how many cells express any given gene? Ordered from rare to common genes. So you might consider focusing on genes that are relatively well expressed across the sample, or you might care about genes expressed only in a small subset of cells. The >=2 refers to read number per cell. 
```{r}
plot(sort(Matrix::rowSums(GetAssayData(sobject)>=2)),xlab="gene rank",ylab="number of cells",main="Cells per genes ( >= 2 )")
```

A. Gene Plot, scatter plot of gene expression across cells, (colored by sample) - sequencing depth relative to genes; looking for similarity across samples. Some cell types can have relatively few genes, but highly expressed, could show up as a second arc below the one seen here (should contain all samples though)
B. % Mitochondrial RNA vs Feature counts
```{r}
Idents(sobject) = "sample_description"
#Save graphs
for (i in 1:(length(m)-1)){
  graph = FeatureScatter(sobject,m[i],m[i+1])+ theme_bw()
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(m[i],"_",m[i+1],project,"scatter.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:(length(m)-1)){
  graph = FeatureScatter(sobject,m[i],m[i+1])+ theme_bw()
  print(graph)
}
```
Set limits based on the graphs
```{r}
mlo = 0.0000
mhi = 10
clo = 1000
chi = 7500
flo = 500 
fhi = 2500
```

Violin plot of 1) number of genes, 2) number of UMI and 3) percent mitochondrial genes; save with lines that show cutoffs

```{r}
group = "orig.ident"
#Make a table with your cutoff values
tablelow = c(clo,flo,mlo)
tablehigh = c(chi,fhi,mhi)
t = as.data.frame(rbind(tablelow,tablehigh))
colnames(t) = m


#save graphs
for (i in 1:ncol(t)){
  metric = colnames(t)[i]
  low = t[1,i]
  high = t[2,i]
  graph = anno_vln(metric = metric) + geom_abline(intercept = high, slope = 0, linetype = "dashed",color = "red", size = 0.8) + geom_abline(intercept = low, slope = 0, linetype = "dashed",color = "black", size = .8)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(colnames(t)[i],project,"withcutoffs.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:ncol(t)){
   metric = colnames(t)[i]
  low = t[1,i]
  high = t[2,i]
  graph = anno_vln(metric = metric) + geom_abline(intercept = high, slope = 0, linetype = "dashed",color = "red", size = 0.8) + geom_abline(intercept = low, slope = 0, linetype = "dashed",color = "black", size = .8)
  print(graph)
}

```



Scatter plots with lines to show cutoffs
```{r}
#Save graphs
for (i in 1:(length(colnames(t))-1)){
  xm = colnames(t)[i]
  ym = colnames(t)[i+1]
  xlow = t[1,i]
  xhigh = t[2,i]
  ylow = t[1,i+1]
  yhigh = t[2,i+1]
  graph = FeatureScatter(sobject,xm,ym)+ theme_bw() + geom_vline(xintercept = xlow, linetype = "dashed",color = "black", size = .8) + geom_vline(xintercept = xhigh, linetype = "dashed",color = "red", size = .8)+ geom_hline(yintercept = ylow, linetype = "dashed",color = "black", size = .8)  + geom_hline(yintercept = yhigh, linetype = "dashed",color = "red", size = .8)
  setEPS()
  postscript(file.path(dir,QCfolder,paste0(m[i],"_",m[i+1],project,"scatter_withcutoffs.eps")))
  print(graph)
  dev.off()
}

#print graphs
for (i in 1:(length(colnames(t))-1)){
  xm = colnames(t)[i]
  ym = colnames(t)[i+1]
  xlow = t[1,i]
  xhigh = t[2,i]
  
  ylow = t[1,i+1]
  yhigh = t[2,i+1]
  
  graph = FeatureScatter(sobject,xm,ym)+ theme_bw() + geom_vline(xintercept = xlow, linetype = "dashed",color = "black", size = .8) + geom_vline(xintercept = xhigh, linetype = "dashed",color = "red", size = .8)+ geom_hline(yintercept = ylow, linetype = "dashed",color = "black", size = .8)  + geom_hline(yintercept = yhigh, linetype = "dashed",color = "red", size = .8)
  print(graph)
}
```
******

Filtering cells: note that they are taking these filters and duplicating them across all samples. Use -Inf instead of 0 because not all the filters have to be positive values, and to avoid any errors due to truncation. 

How to select filters??? noise vs signal - get them from the violin plots and the matrix, capture the bulk of cells but exclude outliers (say with mtDNA filter) - can also look at percentiles, maybe get rid of top 5% since it was very different from the rest. Basically eliminate outliers and don't include cells that have hardly any genes or have a ton of genes expressed. First save the original object: 

```{r}
save(sobject,file = file.path(dir,"outputs",paste0(project,"_beforeQC.RData")))
```

Subset the data. Printing tables lets you see the progression as you tighten your QC cutoffs. 
```{r}
print("initial")
table(sobject$orig.ident) #print out initial cell counts
sobject = subset(sobject, subset = percent.mito>mlo & percent.mito < mhi) #subset mitochondria
print("mito")
table(sobject$orig.ident) #new cell counts
sobject = subset(sobject, subset = nCount_RNA>clo & nCount_RNA < chi)
print("ncount")
table(sobject$orig.ident)
sobject = subset(sobject, subset = nFeature_RNA>flo & nFeature_RNA < chi)
print("nfeature")
table(sobject$orig.ident)
```

```{r}
save(sobject, file = file.path(dir,"outputs",paste0(project,"_filtered.RData")))
```
******

*Edit the following code every time* 
and make sure the folders "QC" and "Plots" and "Heatmaps" and "Trees" and "Spreadsheets" and "Data/Seurat" are present in the "dir" folder
```{r}
#set up folders
QCfolder = "QC"
Plotfolder = "Plots"
datafolder = "outputs"
#the data file you want to open
filename = "hackathon2021-task2_filtered.RData"


#Choose what to regress out - could be age, sex, or any metadata column
regress = c("nCount_RNA","percent.mito")

#Decide whether to use SCtransform (can take a long time with big datasets; generally recommended)
sct = TRUE

#How many genes do you want scaled to use for UMAP/clustering? 
ngenes = 6000

#Which principal components do you want to calculate on? This is a default setting, change if one of the pc's is for something you think is a technical error (i.e. HSP, RP, etc)
pcs = c(1:30)

#clustering resolution; the last number will be saved as "seurat_clusters" in metadata
res = c(1.5,1.0,0.5)

#Important genes to determine your cells of interest
igenes = c("CD19","CD20","CD4","CD8","IL10","CD16","CD56","AIF1")
#CD19,20 - B CELL; CD2,3,4 CD4 T-CELL
#metadata dimensions you want to cluster on
dims = c("seurat_clusters")

#edit to include all desired celltypes to subset on later; if not subsetting, set to "all"
keep = "all"

#make a unique name; maybe this is the celltype you've already subset, or the age you're looking at, etc. 
iterationname = "task_2_sct"

#Establish cutoffs for heatmaps
pval = 1e-8 #max p-value for significance
lfc = 0.2 #minimum log fold change
minpct = 10 #if you want to subset by percent cells in that cluster expressing the gene
maxpct = 1
single = F #should each gene be a marker of a single cluster only
hgenes = 5 #how many genes should be in the heatmap per cluster
ncells = 100 #max # of cells per heatmap column
column = "celltypecluster" #division you care about

#Variance partition: Remove genes only associated with a technical variable or sex of the mouse
variance = F
```


Load in your filtered dataset
```{r}
load(file.path(dir,datafolder,filename))
```

Function to print multiple graphs: 
```{r}
PrintSeuratGraph = function(namecard = "a",seurat_object = sobject,graphtype = "feature",feature = NULL,group = NULL,split=NULL,cellnames=NULL){
  if (!is.null(cellnames)){
    Idents(seurat_object) = cellnames[1]
    cells = colnames(seurat_object)[Idents(seurat_object) %in% cellnames[2:length(cellnames)]]} 
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(seurat_object,features = feature,split.by = split, cells = cells,cols = c("lightyellow","darkred"))
  }
  if (graphtype == "violin"){
    graph = VlnPlot(seurat_object,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(seurat_object,cells = cells, group.by = group, split.by = split)
    
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path(dir,Plotfolder,name))
  print(graph)
  dev.off()
}
```



Find variable features, normalize, scale, run PCA, clustering, umap
The following is the standard method of normalization and scaling. Interchangeable with the next chunk. Run both, you should have already specified which to use with "sct = T or F". Will take 5-20 minutes to run. 

Alternative: SCTransform (great for smaller datasets)
```{r, results='hide',eval = FALSE}
if (sct){
  sobject <- SCTransform(sobject, vars.to.regress = regress, verbose = FALSE,variable.features.n = ngenes,conserve.memory = T)
}
```

******

Show most variable genes
```{r}
labels <- c(head(VariableFeatures(sobject),10))
plot1 = VariableFeaturePlot(sobject)
LabelPoints(plot=plot1, points = labels, repel = F, xnudge = 0.1, ynudge = 0.5)
```

read in the doublet data to see if they've been removed
**note: I need this for individual identification!!

```{r}
dblts = read.table(file.path(dir,"task2.demuxlet.txt"))
head(dblts)
ID = dblts$ANON.ID
names(ID) = rownames(dblts)

ID = ID[match(colnames(sobject),names(ID))]
head(ID)
head(colnames(sobject))

identical(colnames(sobject),names(ID)) #should be true
sobject$id = ID

```

Remove genes that correlate strongly with technical variables from downstream analyses

*****
Run PCA analysis and show elbow plot
```{r}
sobject <- RunPCA(sobject,features = VariableFeatures(sobject),npcs = 50, verbose = FALSE)
ElbowPlot(sobject,ndims = 50, reduction = "pca")
print(sobject[["pca"]], dims = 1:20, nfeatures = 5)
```

Once you are satisfied with pc's, run clustering: 
```{r}
sobject<-RunUMAP(sobject,reduction = "pca",dims = pcs, verbose = F)
sobject<-FindNeighbors(sobject,dims=pcs,verbose=F)
sobject<-FindClusters(sobject,verbose=F,resolution = 1.5)
```

Plot important objects; check parameters before moving forward, evaluate QC, clustering
```{r}
for (dim in dims){
  print(DimPlot(sobject,group.by = dim, label = T))
}

FeaturePlot(sobject,igenes)
FeaturePlot(sobject,m)

VlnPlot(sobject,igenes,sort = "increasing",pt.size = 0.01)

#Build a clustering tree
Idents(sobject) = "seurat_clusters"
sobject= BuildClusterTree(sobject,dims = pcs)
tree = sobject@tools$BuildClusterTree
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
```
****

```{r}
#If you want to see a certain cluster: 
cluster = '3'

#select only the cells from that cluster
cells = sobject$seurat_clusters
cells = cells[cells == cluster]

DimPlot(sobject,cells.highlight = names(cells))
```

If you are happy with the QC, you can move on to either subdividing further to isolate interesting cells (below) or running differential expression comparisons on the existing clustering. 

*Note that the chunk below requires you to manually decide which clusters to keep or exclude. The violin plots will help decide based on marker genes. You can also choose to annotate (as many types as you want) and skip subsetting.*
```{r}
#Edit this part carefully. You can add any number of types. Each cluster can only be one type. 
type1 = c(2,5,7)
name1 = "Astrocytes"
type2 = c(0,4,6,10) 
name2 = "Microglia" 
type3 = c(8)
name3 = "Oligodendrocytes"
type4 = 3
name4 = "Endothelial"
type5 = c(9,1)
name5 = "OPCs"
type6 = c(11,12)
name6 = "Neurons"

type7 = c(14)
name7 = "Pericytes"


#Initialize the cluster levels as a vector and replace the cluster levels with the appropriate name. 
clusters = as.factor(sobject$seurat_clusters)
type = levels(clusters)
type[type1+1] = name1
type[type2+1] = name2
type[type3+1] = name3
type[type4+1] = name4
type[type5+1] = name5
type[type6+1] = name6
type[type7+1] = name7
levels(clusters) = type

#Add a metadata column
sobject$celltype = clusters

#check the celltype assignment for accuracy
table(sobject$celltype,sobject$seurat_clusters)

#Check them against your marker genes
VlnPlot(sobject,igenes,group.by = "celltype",pt.size = 0.01)

#add a metadata column labelling each cluster
sobject$celltypecluster = paste0(sobject$celltype,"-",sobject$seurat_clusters)
```
*****
Save the tree
```{r}
Idents(sobject) = column
sobject= BuildClusterTree(sobject,dims = pcs)
tree = sobject@tools$BuildClusterTree
setEPS()
postscript(file.path(dir,"Trees",paste0(project,"_tree_",iterationname,".eps")))
plot.phylo(tree, use.edge.length = T, direction = "rightwards")
dev.off()
```

Save the clustered dataset. Overwrite the existing, subset dataset. 
```{r,eval = FALSE}
save(sobject,file = file.path(dir,datafolder,filename))
```

Block to print multiple graphs: 
```{r}
name = paste0(project,iterationname)
genes = igenes
features = m
groups = c(dims,"celltype","celltypecluster")
genes = genes[genes %in% rownames(GetAssayData(sobject,slot = "data"))]

for(feature in genes){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

for(feature in features){
  PrintSeuratGraph(namecard = name,graphtype = "feature",feature = feature)
}

#split feature plots by individual
for(feature in c(features)){
  PrintSeuratGraph(namecard = paste0(name,"_split"),graphtype = "feature",feature = feature,split = "sample_description")
}

#dim plots for clustering
for(group in groups){
  PrintSeuratGraph(namecard = name,graphtype = "dim",group = group, feature = group)
}

#violin plots
for(feature in c(genes,features)){
  PrintSeuratGraph(namecard = name,graphtype = "violin",feature = feature,group = "seurat_clusters")
}
```

*Heatmap*

First calculate DE genes for every cluster
```{r,results = "hide",eval = FALSE}
Idents(sobject) = column
markers_all <- FindAllMarkers(
    object = sobject,
    features = VariableFeatures(sobject),
    test.use = "MAST",
    only.pos = FALSE, 
    min.pct = 0.05, 
    logfc.threshold = 0.0
)
```


```{r,results = "hide",eval = FALSE}
write.csv(markers_all,file = file.path(dir,"Spreadsheets",paste0(iterationname,"_all_markers.csv")))
```

Make the heatmap
```{r}
#read in a de gene file
markers = read.csv(file.path(dir,"Spreadsheets",paste0(iterationname,"_all_markers.csv")),stringsAsFactors = F)

#Select only the genes that pass thresholds
markers = markers[markers$p_val_adj<pval,]

#pick only positives, or restrict by min/max pct expression using pct1/2
markers = markers[markers$avg_logFC > lfc,] 
markers = markers[markers$pct.1 > minpct & markers$pct.2 < maxpct,]

#If you want, select markers that define a single cluster
if (single){markers <- markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) == 1],] }

table(markers$cluster)

topgenes <- markers %>% group_by(cluster) %>% top_n(ngenes, avg_logFC)
topgenes = topgenes[order(topgenes$cluster),]

#Subset each cluster to ncells
cellnames = sobject@meta.data[,column]
names(cellnames) = colnames(sobject)
clusters = levels(as.factor(cellnames))
newcellnames = NULL
for (cluster in clusters){
  n = length(cellnames[cellnames == cluster])
  if (n > ncells){n = ncells}
  newcluster = sample(cellnames[cellnames == cluster],n, replace = F)
  newcellnames = c(newcellnames,newcluster)
}

#check
table(newcellnames)

#Make heatmap
setEPS()
postscript(file.path(dir,"Heatmaps", paste0(iterationname,"_",column,"_",pval,"_ncells",ncells,"heatmap.eps")))
DoHeatmap(
    object = sobject, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()
```

Print heatmap to console

```{r}
DoHeatmap(
    object = sobject, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 5,
    label = T,
    draw.lines = T
)
```

Subset the data if necessary; will only run if "keep" includes some celltype

Subset the data to include only your celltypes of interest: 
```{r}
if (all(keep %in% sobject$celltype)){
  sobject = subset(sobject,subset = celltype %in% keep)
  #check the resulting subset
  table(sobject$celltype,sobject$seurat_clusters)
  #subset the object by the metadata column "celltype"
  save(sobject,file = file.path(dir,datafolder,paste0(project,"_",keep,"_subset.RData")))
}
```

Conduct differential expression analysis on the celltypes (pooled)
```{r}
column = "celltype"
ncells = 500
```

```{r,eval = FALSE}
Idents(sobject) = column
markers_all = FindAllMarkers(
  object = sobject,
  features = rownames(sobject),
  test.use = "MAST", 
  only.pos = FALSE, 
  min.pct = 0.15, 
  logfc.threshold = 0.0)
write.csv(markers_all,file = file.path(dir,"Spreadsheets",paste0(iterationname,"_",column,"_markers.csv")))
```

```{r}
#read in a de gene file
markers = read.csv(file.path(dir,"Spreadsheets",paste0(iterationname,"_",column,"_markers.csv")),stringsAsFactors = F)

#Select only the genes that pass thresholds
markers = markers[markers$p_val_adj<pval,]

#pick only positives, or restrict by min/max pct expression using pct1/2
markers = markers[markers$avg_logFC > lfc,] 
markers = markers[markers$pct.1 > minpct & markers$pct.2 < maxpct,]

#If you want, select markers that define a single cluster
if (single){markers <- markers[markers$gene %in% names(table(markers$gene))[table(markers$gene) == 1],] }

table(markers$cluster)

topgenes <- markers %>% group_by(cluster) %>% top_n(hgenes, avg_logFC)
topgenes = topgenes[order(topgenes$cluster),]

#Subset each cluster to ncells
cellnames = sobject@meta.data[,column]
names(cellnames) = colnames(sobject)
clusters = levels(as.factor(cellnames))
newcellnames = NULL
for (cluster in clusters){
  n = length(cellnames[cellnames == cluster])
  if (n > ncells){n = ncells}
  newcluster = sample(cellnames[cellnames == cluster],n, replace = F)
  newcellnames = c(newcellnames,newcluster)
}

#check
table(newcellnames)

#Make heatmap
setEPS()
postscript(file.path(dir,"Heatmaps", paste0(iterationname,"_",column,"_",pval,"_ncells",ncells,"heatmap.eps")))
DoHeatmap(
    object = sobject, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 5,
    label = T,
    draw.lines = T
)

dev.off()
```

Print heatmap to console

```{r}
DoHeatmap(
    object = sobject, 
    features = c(topgenes$gene),
    cells = names(newcellnames),
    group.by = column,
    size = 5,
    label = T,
    draw.lines = T
)
```
****