---
title: "adultjuvenileoverlay"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

---
title: "bulksinglecelloverlay"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

Calculate a module eigengene

load in the single cell object
```{r}
load("~/Desktop/Sequencing/Nick/2_2020/Graphs_Analysis_August2020 mpeg only/Data/mpeg_adult.RData")
```

```{r}
PrintSeuratGraph = function(namecard = "a",sobject,graphtype = "feature",colors = c("lightgrey","lightblue","yellow","orange","darkorange","red"), feature = NULL,group = NULL,split=NULL,cellnames=NULL,sort = F){
  if (!is.null(cellnames)){
    Idents(sobject) = cellnames[1]
    cells = colnames(sobject)[Idents(sobject) %in% cellnames[2:length(cellnames)]]} 
  else {cells = cellnames}
  if (graphtype == "feature"){
    graph = FeaturePlot(sobject,features = feature,split.by = split, cells = cells,cols = colors)
  }
  if (graphtype == "violin"){
    graph = VlnPlot(sobject,features = feature, pt.size = 0.1, idents = cellnames[2:length(cellnames)],group.by = group, split.by = split)
  }
  if (graphtype == "dim"){
    graph = DimPlot(sobject,cells = cells, group.by = group, split.by = split)
    
  }
  name = paste0(feature,"_",graphtype,namecard,".eps")
  graph
  setEPS()
  postscript(file.path("~/Desktop",name))
  print(graph)
  dev.off()
}

```

Read in DE gene list for juvenile clusters
```{r}
juvenile = read.csv("~/Desktop/Sequencing/Nick/2_2020/Graphs_Analysis_August2020 mpeg only/New analyses - updated QC/Juvenile_withoutcluster1/Spreadsheets/juvenile_mpeg_markers.csv")
```

Create consensus lists for each region

```{r}
pval = 1e-10
lfc = 0.21 #1.5

juvenile <- juvenile[juvenile$gene %in% names(table(juvenile$gene))[table(juvenile$gene) == 1],] #markers that define a single cluster
juvenile = juvenile[juvenile$p_val_adj<pval,]
juvenile = juvenile[juvenile$avg_logFC > lfc,]
l = list()
juvenile$cluster = as.factor(juvenile$cluster)
levels(juvenile$cluster)
for (cluster in levels(juvenile$cluster)){
  l[[cluster]] = juvenile[juvenile$cluster == cluster,]
}
l
  
```



```{r}
for (i in 1:length(l)){
  genes = l[[i]]$gene
  sobject = PercentageFeatureSet(sobject,features = genes,col.name = paste0("juvcluster_",names(l)[i]))
}
```


```{r}
features = c("juvcluster_0","juvcluster_1","juvcluster_2","juvcluster_3","juvcluster_4")

for(feature in features){
  PrintSeuratGraph(namecard = "adult-green",sobject=sobject,graphtype = "feature",feature = feature,colors = c("light grey","lightgreen","forestgreen","darkgreen","black"))
}


for(feature in features){
  PrintSeuratGraph(namecard = "adult-blue",sobject=sobject,graphtype = "feature",feature = feature,colors = c("light grey","lightblue","blue","darkblue","black"))
}

for(feature in features){
  PrintSeuratGraph(namecard = "adult-red",sobject=sobject,graphtype = "feature",feature = feature,colors = c("light grey","red","maroon","darkred","black"))
}

for(feature in features){
  PrintSeuratGraph(namecard = "adult",sobject=sobject,graphtype = "violin",feature = feature)
}

PrintSeuratGraph(namecard = "adult",sobject = sobject,graphtype = "dim",feature = "seurat_clusters")
```

